Decision: Spec Expressiveness (#4743)
ElektraInitiative/libelektra: doc/decisions/1_in_discussion/spec_expressiveness.md

## Problem

The current implementation of the spec plugin is copying metadata from the specification

> (`spec` namespace) to all other namespaces. By providing certain metadata e.g. 

Reviewer: Please use line breaks as described in doc/contrib/documentation.md

`type` or `description`, it can be assured that keys adhere to a given specification. 

For some applications there are keys we don't want to define for every application explicitly.
This is when `globbing` expressions in key names comes into place. By providing certain wildcard
characters e.g. `_` or `#` one can match and specify a number of keys at once without the need
to explicitly state the key in every specification.

To describe the use case where `globbing` expressions make sense, we will use an application

> based on microservice architecture as sample.

Reviewer: Please describe use-cases separately (in another PR). For web-UI there is already doc/usecases/elektra_web/UC_key_validation.md

We also need use-cases directly for the spec lang, e.g., for `#` and `_` (we probably don't need `*`).

Imagine we have microservices and each of them communicates to a database to persist data. We
will need some configuration so the applications can connect to the database e.g. `port`. Lets
say we are specifying our configuration by `database/<DIALECT>/...`, where `...` is any name
for a key. If we have multiple dialects e.g. a microservice uses more than one database e.g. `mysql`, `mssql`, etc. it is inconvenient to copy and paste the configuration for every 
dialect we use.

By using `globbing` expression we can overcome this problem.

The key starts with `/sw/org/app-name`.
```ni
[database/_/port]
meta:/require = true
meta:/description = "The database port."
```

> The `_` would match every `dialect` we are using in our application e.g. `postgresql`, `mssql`, etc.

Reviewer 2: The example is a bit odd IMO, because it only makes sense, if one instance of the application uses multiple databases at the same time. If it just needs a single database, but lets the user choose which database the want to use, the specification could just be:

```ni
[database/port]
meta:/require = true
meta:/description = "The database port."

[database/dialect]
meta:/require = true
meta:/description = "The database dialect."
```

Maybe a better example, would be a service that acts as a reverse proxy:

```ni
[proxy/_/url]
meta:/require = true

[proxy/domain]
meta:/require = true
```

So this config

```
proxy/domain = example.com

proxy/bing/url = bing.com
proxy/google/url = google.com
```

would proxy requests for `bing.example.com` to `bing.com` and requests for `google.example.com` to `google.com`.

Or another example would be a mail client, which supports checking multiple accounts:

```ni
[account/#/email]
meta:/require = true

[account/#/password]
meta:/require = true
```

## Problem

Specifications can help users define certain details about keys and make sure only valid keys are used.

One example of such as specification can be found below:

```ni
[public/ip]
meta:/default = 127.0.0.1
meta:/description = "The IP via which this application is accessible."

[public/port]
meta:/default = 8080
meta:/description = "The port on which this application is accessible."
```

Here one can see `default` and `description` used as metadata. 

The current implementation of the spec plugin is copying this metadata to all other namespaces.

> It also assures that if keys do not exist and have `default` metadata, it creates a cascading key.

Reviewer: ```suggestion
It also assures that if keys do not exist and have `default` metadata, it creates a key in the default namespace.
```

Author: Thanks!

> One example of such as specification can be found below:
> 
> ```ni
> [public/ip]
> meta:/default = 127.0.0.1
> meta:/description = "The IP via which this application is accessible."
> 
> [public/port]
> meta:/default = 8080
> meta:/description = "The port on which this application is accessible."
> ```
> 
> Here one can see `default` and `description` used as metadata. 
> 
> The current implementation of the spec plugin is copying this metadata to all other namespaces.
> It also assures that if keys do not exist and have `default` metadata, it creates a cascading key.
> 
> Assume we have a REST application we want to write a specification for.
> This REST application has to use multiple services.

Reviewer: What is this first example for?

```suggestion
```

Author: I started this decision to first explain how something works and then slowly going towards to actual problem so everyone understands why this problem exists.

If we have `database`, `broker` and `mailer` services we could use the specification:

```ni
[service/database/port]
meta:/require = true
meta:/default = 8080
meta:/description = "The port of the database service."

[service/database/url]
meta:/require = true
meta:/description = "The url of the database service."

[service/broker/port]
meta:/require = true
meta:/default = 8080
meta:/description = "The port of the broker service."

[service/broker/url]
meta:/require = true
meta:/description = "The url of the broker service."

[service/mailer/port]
meta:/require = true
meta:/default = 8080
meta:/description = "The port of the mailer service."

[service/mailer/url]
meta:/require = true
meta:/description = "The url of the mailer service."
```

This is already very repetitive and error prone. 
But writing the specification becomes completely impossible when the set of services is not known in advance. 
To solve this we need some way to allow for dynamic/unknown parts in the keyname.

> `Globbing` expressions is the solution to this problem. 

Reviewer: The problem part should only contain the problem.

Author: I already have just written down the problems in a previous version. I was told to back it up with some example, which was a very good input IMO, as I think people reading this document immediately understand what this problem is about without the need of knowing to much details.

By providing certain wildcard characters e.g. `_` or `#` one can match multiple keys.
This removes the need to explicitly define specification for keys in our example for every service. 

The thing to be aware of is to define configuration keys which are necessary for every service.
In this example we use `port` and `url`.

The parent key for our specification is `/sw/org/app-name/#0/development`.

```ni
[service/_/port]
meta:/require = true
meta:/default = 8080
meta:/description = "The port of the service."

[service/_/url]
meta:/require = true
meta:/description = "The url of the service."
```
The `_` would match every `service` we are using in our application e.g. `database`, `message broker`, etc.

> But the current implementation has some undefined behaivour when using `globbing`.

Reviewer: I think the actual problem starts here.

```suggestion
A naive  implementation has some undefined behavior when using `globbing`.
```

This specification uses `_` as wildcard character.
Unfortunately `_` has a shortcoming, as we do not know the key names in advance.
This prevents us from creating those keys in namespace they do not exist or copying metadata in advance.

But besides `_` we also have `#` as wildcard character.
This wildcard character is used with arrays.
A fixed array size can be defined in the specification, which overcomes the shortcoming of `_`.

The problem we face with those wildcard characters is overlapping specifications.
A example for this is if we used `[service/#/port]` and `[service/shtg/port]` or `[service/_/port]` in the same specification.
We don't know what metadata should be copied to other namespaces in this case.

> Both `[service/#/port]` and `[service/_/port]` existing produces conflicting specification.

Reviewer: existing?

Author: Thanks!

For example `[service/#/port]` having as type `meta:/type = string`.
`[service/shtg/port]` having `meta:/type = unsigned_short`.

> The spec plugin can not know which of those specification to use.

Reviewer: Again: this is already solution.

Btw. Of course some behavior could be defined, so the spec plugin can know what to use.

Author: > Btw. Of course some behavior could be defined, so the spec plugin can know what to use.

So you are saying that this is not a problem at all? Having both `[service/#/port]` and `[service/shtg/port]` with different `meta:/type` can be configured the way it works?

## Problem

The current implementation of the spec plugin is copying metadata to all other namespaces.
Metadata e.g. `type` or `description` describes a certain key.

Sometimes there are keys with similar or identical specifications.
We want to define those in a single place and not have to manually copy metadata.
`Globbing` expressions in key names comes into place at that time. 
By providing certain wildcard characters e.g. `_` or `#` one can match multiple keys.

> This removes the need to explicitly define specification for keys.

Reviewer 2: Weird order. I'd start with the "Why?" then explain the "How?". So first give an example of specifications and then explain that the current `spec` achieves this by copying metadata.

You should probably also start this without mentioning globbing. A specification without globbing still makes sense and is easier to understand. In your example you could for example have

```ni
[public/ip]
meta:/default = 127.0.0.1
meta:/description = "The IP via which this application is accessible."

[public/port]
meta:/default = 8080
meta:/description = "The port on which this application is accessible."
```

To describe the use case where `globbing` expressions make sense, we will use a REST application.
This REST application uses multiple services which the user is not aware of at time of configuration.
It could use a database, message broker, etc. but this is not yet known.
A solution to this is `globbing` i.e. using wildcards in key names.
The thing to be aware of is to define configuration keys which are necessary for every service.
In this example we use `port` and `url`.

The parent key for our specification is `/sw/org/app-name/#0/development`.

```ni
[service/_/port]
meta:/require = true
meta:/default = 8080
meta:/description = "The port of the service."

[service/_/url]
meta:/require = true
meta:/description = "The url of the service."
```

> The `_` would match every `service` we are using in our application e.g. `database`, `message broker`, etc.

Reviewer 2: If I want to explain the need for globbing IMO it is best to show that it is worse (or impossible) without globbing. In your example you could start like this:

````markdown
If we have `database`, `broker` and `mailer` services we could use the specification:

```ni
[service/database/port]
meta:/require = true
meta:/default = 8080
meta:/description = "The port of the database service."

[service/database/url]
meta:/require = true
meta:/description = "The url of the database service."

[service/broker/port]
meta:/require = true
meta:/default = 8080
meta:/description = "The port of the broker service."

[service/broker/url]
meta:/require = true
meta:/description = "The url of the broker service."

[service/mailer/port]
meta:/require = true
meta:/default = 8080
meta:/description = "The port of the mailer service."

[service/mailer/url]
meta:/require = true
meta:/description = "The url of the mailer service."
```

This is already very repetitive and error prone. But writing the specification becomes completely when the set of services is not known in advance. To solve this we need some way to allow for dynamic/unknown parts in the keyname. That is what we call Globbing.

[...]
````

Author: Thank you, this is much better.

But the current implementation has some undefined behaivour when using `globbing`.

> This specification uses `_` but we also have `#` as wildcard in `globbing`. 

Reviewer 2: I would first explain the issues with `_` alone, i.e. that we cannot know the keynames that will be used with `ksLookup`, so we cannot create keys or copy metadata in advance.

Then you could mention that `#` also exists, and that because it is restricted to arrays it does _not_ have this problem. We only need to know the array size to create keys and copy metadata. We can simply enforce that the array size is defined.

Only then would I mentioned that `#` doesn't solve all problems. Overlapping specification is still and issue, but that's easy to fix. I'd just mention that any solution must report errors for overlapping specifications, something the current `spec` fails to do.

Finally, you should also mention the disadvantages of using `#` over `_`. Take your example, but with `#`:

```ni
[service/#/port]
meta:/require = true
meta:/default = 8080
meta:/description = "The port of the service."

[service/#/url]
meta:/require = true
meta:/description = "The url of the service."
```

Here we loose the unique name for the service that was previously covered by the `_` part. Now we just get an index in the array. Still a unique identifier, but much harder to handle. Also it's not easily possible to give meaning to this index. Whereas before using `database` for `_` could indicate something to the application (and the user), you now only have something like `service/#0/port`. You can't really give meaning to `#0`.

That's why you'll often end up with things like:

```ni
[service/#/port]
meta:/require = true
meta:/default = 8080
meta:/description = "The port of the service."

[service/#/url]
meta:/require = true
meta:/description = "The url of the service."

[service/#/name]
meta:/default = ""
meta:/description = "The name of the service."
```

Now we have a name again. But that is still more annoying to use. You have to set more keys. The key `service/#0/url` still doesn't tell you much. You still have to look at `service/#0/name` to find out what the URL is for, etc.

That is why we would like to have a `_` that works at least somewhat decently.

> **Note**: Using `service/#/name` to define what kind of service this array entry is, also has an advantage over using the `_` part for this. The value of `service/#/name` can be restricted with a `meta:/check/enum`, while `_` is completely arbitrary. The tags idea in #3954 could address that.

## Problem

The current implementation of the spec plugin is copying metadata to all other namespaces.
Metadata e.g. `type` or `description` describes a certain key.

Sometimes there are keys with similar or identical specifications.
We want to define those in a single place and not have to manually copy metadata.
`Globbing` expressions in key names comes into place at that time. 
By providing certain wildcard characters e.g. `_` or `#` one can match multiple keys.
This removes the need to explicitly define specification for keys.

To describe the use case where `globbing` expressions make sense, we will use a REST application.
This REST application uses multiple services which the user is not aware of at time of configuration.
It could use a database, message broker, etc. but this is not yet known.

> A solution to this is `globbing` i.o. using wildcards in key names.

Author: Done.

The thing to be aware of is to define configuration keys which are necessary for every service.
In this example we use `port` and `url`.

The key starts with `/sw/org/app-name/#0/development`.
`sw` defines that this is a configuration for a software.
`org` is the organization name.
`app-name` is the application name we are writing the specification for.
`#0` is the configuration version.

> `development` is the profile (it could also be `production`) 

Author: Ok, I understand your point. It thought it would be good to add in case anyone comes across and does not remember the convention anymore.

> `development` is the profile (it could also be `production`) 

Author: Done.

```ni
[service/_/port]
meta:/require = true
meta:/default = 8080
meta:/description = "The port of the service."

[service/_/url]
meta:/require = true
meta:/description = "The url of the service."
```
The `_` would match every `service` we are using in our application e.g. `database`, `message broker`, etc.

### Problem 1

This specification uses `_` but we also have `#` as wildcard in `globbing`. 
The problem we face here is overlapping specifications.
A example for this is if we used `[service/#/port]` and `[service/_/port]` in the same specification.

> This is currently undefined behaivour in the spec plugin.

Reviewer: There can also be `service/sthg/port`.

Author: yes, thanks.

Author: Done.

### Problem 2

> Errors and warnings are strictly defined in elektra.

Reviewer: ```suggestion
Errors and warnings are strictly defined in Elektra.
```

Author: Done.

> The spec plugin is currently not 

Reviewer 2: Incomplete sentence

### Problem 3

> Default values on wildcard globbing is undefined.

Reviewer 2: What does that mean?

## Decision

- no defaults for keys ending with globbing keys `sw/_` (**Problem 3**)
- no defaults for `sw/_/keys` (default will not work for `ksLookup(/sw/sthg/key)`)
- The spec plugin should yield should yield errors to user on overlapping keys (**Problem 1**)

> - The spec plugin should yield errors and warnings when it detects such situations (**Problem 2**)

Author: Yes, exactly.
